src/client_side_request.cc

```diff
@@ -614,14 +615,17 @@ ClientRequestContext::hostHeaderVerify()
         // Verify forward-proxy requested URL domain matches the Host: header
         debugs(85, 3, "FAIL on validate URL domain " << http->request->url.host() << " matches Host: " << host);
         hostHeaderVerifyFailed(host, http->request->url.host());
-    } else if (portStr && port != http->request->url.port()) {
+    } else if (portStr && !http->request->url.port()) {
+        debugs(85, 3, "FAIL on validate portless URI matches Host: " << portStr);
+        hostHeaderVerifyFailed("portless URI", portStr);
+    } else if (portStr && port != *http->request->url.port()) {
         // Verify forward-proxy requested URL domain matches the Host: header
```

Comment:
The new `if` statement above (line 618) is needed so that we can safely dereference url.port() in the other `if` statement on line 621 and 623. AFAICT, this change preserves the original logic.

The official code here is rather messy/problematic and should be refactored, but this PR is not the right place to do that. If you want debugs() or hostHeaderVerifyFailed() parameters to say something else, please provide exact strings (while keeping the peculiarities of the surrounding official code in mind). Thank you.

Reply:
I've spent a while looking this over now and does seem to be both correct and the cleanest current change.

FTR, status quo is what it is because of the incremental discovery of the "safe" or "only partially unsafe" cases in the sparse matrix of input permutations. A refactor risks obscuring which (and how) cases are handled, do not attempt expecting an easy or quick improvement.